### PR TITLE
8340383: VM issues warning failure to find kernel32.dll on Windows nanoserver

### PR DESCRIPTION
**Pull request by gdams**
On Windows Nanoserver the `kernel32.dll` is not within the System32 directory itself but within the "forwarders" sub-folder.

See a similar bug reported in dotnet a while back: https://github.com/dotnet/sdk/issues/6056

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8340383](https://bugs.openjdk.org/browse/JDK-8340383): VM issues warning failure to find kernel32.dll on Windows nanoserver (**Bug** - P3)

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Julian Waters](https://openjdk.org/census#jwaters) (@TheShermanTanker - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21070/head:pull/21070` \
`$ git checkout pull/21070`

Update a local copy of the PR: \
`$ git checkout pull/21070` \
`$ git pull https://git.openjdk.org/jdk.git pull/21070/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21070`

View PR using the GUI difftool: \
`$ git pr show -t 21070`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21070.diff">https://git.openjdk.org/jdk/pull/21070.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21070#issuecomment-2358989161)

### CODE REVIEW

File: src/hotspot/os/windows/os_windows.cpp
Comment (karianna) [inline src/hotspot/os/windows/os_windows.cpp:4112]: Should this detection logic be extracted for other scenarios wher we need to special case nano?

**Review by karianna — COMMENTED** [review]
detection logic extraction

**Review by dholmes-ora — COMMENTED** [review]
Executing in a container environment (per linux experience) can affect a lot of things in the VM. What exactly is NanoServer and is it a supported environment for OpenJDK? The current PR seems a point fix, but do we need to update other logic - e.g. version recognition to identify and report that we are running on NanoServer?

Thanks

File: src/hotspot/os/windows/os_windows.cpp
Comment (gdams) [reply src/hotspot/os/windows/os_windows.cpp]: I've found a registry key that we can reliably use to detect Windows Nanoserver

**Comment by gdams** [comment]
> Executing in a container environment (per linux experience) can affect a lot of things in the VM. What exactly is NanoServer and is it a supported environment for OpenJDK? The current PR seems a point fix, but do we need to update other logic - e.g. version recognition to identify and report that we are running on NanoServer?
> 
> Thanks

Agreed, noting that previous OpenJDK versions have worked just fine in a Nanoserver environment. This was broken by the newly introduced dependency on kernel32.dll. I've refactored the code now to check the registry to determine if the host OS is a nanoserver variant. This will mean that we can rely on the isNanoServer check in other places in the future where required. As things stand this is enough to get Java working again

File: src/hotspot/os/windows/os_windows.cpp
Comment (karianna) [inline src/hotspot/os/windows/os_windows.cpp:4102]: Do we want to log when we can't read this value?

File: src/hotspot/os/windows/os_windows.cpp
Comment (karianna) [inline src/hotspot/os/windows/os_windows.cpp:4118]: Not sure how OpenJDK source code typically deals with magic strings, but should this get extracted to a const?

File: src/hotspot/os/windows/os_windows.cpp
Comment (TheShermanTanker) [inline src/hotspot/os/windows/os_windows.cpp:4113]: HotSpot code is not allowed to use anything from the C++ Standard Library, aside from very few exceptions. <string> is unfortunately not one of those exceptions

Comment (gdams) [reply src/hotspot/os/windows/os_windows.cpp]: thanks for the heads up, I've refactored this code to remove the dependency on <string> if you wouldn't mind taking another look?

File: src/hotspot/os/windows/os_windows.cpp
Comment (dholmes-ora) [inline src/hotspot/os/windows/os_windows.cpp:4093]: Nit: we don't use this naming style in hotspot. This should be either `getWindowsInstallationType` or `get_windows_installation_type`. Thanks

File: src/hotspot/os/windows/os_windows.cpp
Comment (dholmes-ora) [inline src/hotspot/os/windows/os_windows.cpp:4104]: Is this a protected key or should we generally expect reading it to always succeed?

File: src/hotspot/os/windows/os_windows.cpp
Comment (dholmes-ora) [inline src/hotspot/os/windows/os_windows.cpp:4119]: Again naming style.

File: src/hotspot/os/windows/os_windows.cpp
Comment (dholmes-ora) [inline src/hotspot/os/windows/os_windows.cpp:4123]: Just curious do we really need `lstrcmpA` here rather than plain old `strcmp`?

File: src/hotspot/os/windows/os_windows.cpp
Comment (gdams) [reply src/hotspot/os/windows/os_windows.cpp:4104]: it is a protected key but we're only opening it in read mode. I've tested this with non-elevated users and it seems to run okay

File: src/hotspot/os/windows/os_windows.cpp
Comment (gdams) [reply src/hotspot/os/windows/os_windows.cpp]: updated

File: src/hotspot/os/windows/os_windows.cpp
Comment (gdams) [reply src/hotspot/os/windows/os_windows.cpp]: updated

File: src/hotspot/os/windows/os_windows.cpp
Comment (gdams) [reply src/hotspot/os/windows/os_windows.cpp]: strcmp is fine, I've switched it

**Comment by TheShermanTanker** [comment]
Looks ok to me for the most part. Will approve once David does

**Review by dholmes-ora — APPROVED** [review]
Thanks for the updates. Seems okay.

**Review by TheShermanTanker — APPROVED** [review]

**Comment by gdams** [comment]
would someone please sponsor this change :) thank you

**Comment by TheShermanTanker** [comment]
> would someone please sponsor this change :) thank you

I almost did, but remembered the 24 hour rule. Is this trivial enough to integrate now?

**Comment by gdams** [comment]
> > would someone please sponsor this change :) thank you
> 
> I almost did, but remembered the 24 hour rule. Is this trivial enough to integrate now?

I would have said so and it fixes a breaking change but I'll leave it up to others to decide

**Comment by dholmes-ora** [comment]
@TheShermanTanker  the 24-hour rule is from the creation of the PR to the integration. ???

/sponsor

**Pull request closed without merging**
